OrbitLinuxTracing/UprobesReturnAddressManager.h

```diff
@@ -59,6 +61,74 @@ class UprobesReturnAddressManager {
     }
   }
 
+  // In case of callchain sampling we don't have the complete stack to patch,
+  // but only the callchain (as list of instruction pointers). In those,
+  // a uprobe address occurs in place of the caller of an instrumented function.
+  // This function patches the callchain, using the maps information to identify
+  // instruction pointers of uprobe code and using the return address saved in
+  // the uprobes.
+  bool PatchCallchain(pid_t tid, uint64_t* callchain, uint64_t callchain_size,
+                      unwindstack::Maps* maps) {
+    if (!tid_uprobes_stacks_.contains(tid)) {
+      return true;
+    }
+
+    auto& tid_uprobes_stack = tid_uprobes_stacks_.at(tid);
+    CHECK(!tid_uprobes_stack.empty());
+
+    std::vector<uint64_t> frames_to_patch;
+
+    // TODO(kuebler): What about tail-call optimization, where two uretprobes
+    //  hijacked an address at the same stack pointer.
+    for (uint64_t i = 0; i < callchain_size; i++) {
+      uint64_t ip = callchain[i];
+      unwindstack::MapInfo* map_info = maps->Find(ip);
+
+      if (map_info == nullptr || map_info->name != "[uprobes]") {
+        continue;
+      }
+
+      frames_to_patch.push_back(i);
+    }
+
+    // In case we already used all uprobes, we need to discard this sample.
+    // There are two situations where this may happen:
+    //  1. At the beginning of a capture, where we missed the first uprobes
+    //  2. When some events are lost or processed out of order.
+    if (tid_uprobes_stack.size() < frames_to_patch.size()) {
+      ERROR(
+          "Discarding sample in a uprobe as some uprobe records are missing.");
+      return false;
+    }
+    // In cases of lost events, or out of order processing, there might be wrong
```

Comment:
Are lost events or out of order processing the causes of `tid_uprobes_stack.size() > frames_to_patch.size() + 1`? This seems more like a case that shouldn't happen and that if/when does happen the reason could be unclear.

Reply:
Well, what happen for me was, that there was new uprobes before the uretprobes of the previous uprobe.

Reply:
Ah, makes sense, so basically when a uretprobe is lost or late. (This is an unfortunate case that hopefully doesn't happen on release builds because if it caused exactly `tid_uprobes_stack.size() > frames_to_patch.size() + 1` then it would fall in the other case and would be handled incorrectly.)